controllers/nifi_controller.go

```diff
@@ -0,0 +1,1092 @@
+		if err := controllerutil.SetControllerReference(instance, configMap, r.Scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Check if this ConfigMap already exists
+		existingConfigMap := &corev1.ConfigMap{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}, existingConfigMap); err != nil {
+			if errors.IsNotFound(err) {
+				rlog.Info("Creating a new ConfigMap", "Namespace", configMap.Namespace, "Name", configMap.Name)
+				err = r.Client.Create(context.TODO(), configMap)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		} else {
+			changed := false
+			if !reflect.DeepEqual(configMap.Data, existingConfigMap.Data) {
+				existingConfigMap.Data = configMap.Data
+				changed = true
+			}
+
+			if changed {
+				rlog.Info("Reconciling existing ConfigMap", "Namespace", configMap.Namespace, "Name", configMap.Name)
+				err = r.Client.Update(context.TODO(), existingConfigMap)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		}
+	}
+
+	// Reconcile Nifi Secret
+	{
+		secret := newNifiSecret(nifiNamespacedName)
+		if err := controllerutil.SetControllerReference(instance, secret, r.Scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Check if this secret already exists
+		existingSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: secret.Name, Namespace: secret.Namespace}, existingSecret); err != nil {
+			if errors.IsNotFound(err) {
+				rlog.Info("Creating a new Secret", "Namespace", secret.Namespace, "Name", secret.Name)
+				err = r.Client.Create(context.TODO(), secret)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		} else {
+			changed := false
+			if !reflect.DeepEqual(secret.Data, existingSecret.Data) {
+				existingSecret.Data = secret.Data
+				changed = true
+			}
+
+			if changed {
+				rlog.Info("Reconciling existing Secret", "Namespace", secret.Namespace, "Name", secret.Name)
+				err = r.Client.Update(context.TODO(), existingSecret)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// newServiceAccount returns a new ServiceAccount for a Nifi instance
+func newServiceAccount(meta types.NamespacedName) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nifiPrefix + meta.Name,
+			Namespace: meta.Namespace,
+		},
+	}
+}
+
+// newNifiStatefulSet returns an updated instance of the StatefulSet for deploying Nifi
+func newNifiStatefulSet(ns types.NamespacedName, instance *bigdatav1alpha1.Nifi) *appsv1.StatefulSet {
+	// Image
+	var image string
+	if instance.Spec.Image != "" {
+		image = instance.Spec.Image
+	} else {
+		image = nifiImageRepo + nifiVersion
+	}
+
+	// Replicas
+	var replicas int32
+	if instance.Spec.Size != 0 {
+		replicas = instance.Spec.Size
+	} else {
+		replicas = 1
+	}
```

Comment:
I don't understand what you are doing here, why not just set the replicas to the number provided by the user? 0 is a valid replica count.